runtime/parachains/src/disputes.rs

```diff
@@ -0,0 +1,1963 @@
+					&validator_public,
+					set.candidate_hash,
+					set.session,
+					statement,
+					signature,
+				).map_err(|()| Error::<T>::InvalidSignature)?;
+
+				let valid = match statement {
+					DisputeStatement::Valid(_) => true,
+					DisputeStatement::Invalid(_) => false,
+				};
+
+				importer.import(*validator_index, valid).map_err(Error::<T>::from)?;
+			}
+
+			importer.finish()
+		};
+
+		// Apply spam slot changes. Bail early if too many occupied.
+		let is_local = <Included<T>>::contains_key(&set.session, &set.candidate_hash);
+		if !is_local {
+			let mut spam_slots: Vec<u32> = SpamSlots::<T>::get(&set.session)
+				.unwrap_or_else(|| vec![0; n_validators]);
+
+			for (validator_index, spam_slot_change) in summary.spam_slot_changes {
+				let spam_slot = spam_slots.get_mut(validator_index.0 as usize)
+					.expect("index is in-bounds, as checked above; qed");
+
+				match spam_slot_change {
+					SpamSlotChange::Inc => {
+						ensure!(
+							*spam_slot < config.dispute_max_spam_slots,
+							Error::<T>::PotentialSpam,
+						);
+
+						*spam_slot += 1;
+					}
+					SpamSlotChange::Dec => {
+						*spam_slot = spam_slot.saturating_sub(1);
+					}
+				}
+			}
+
+			SpamSlots::<T>::insert(&set.session, spam_slots);
+		}
+
+		if fresh {
+			Self::deposit_event(Event::DisputeInitiated(
+				set.candidate_hash,
+				if is_local { DisputeLocation::Local } else { DisputeLocation::Remote },
+			));
+		}
+
+		{
+			if summary.new_flags.contains(DisputeStateFlags::FOR_SUPERMAJORITY) {
+				Self::deposit_event(Event::DisputeConcluded(
+					set.candidate_hash,
+					DisputeResult::Valid,
+				));
+			}
+
+			// It is possible, although unexpected, for a dispute to conclude twice.
+			// This would require f+1 validators to vote in both directions.
+			// A dispute cannot conclude more than once in each direction.
+
+			if summary.new_flags.contains(DisputeStateFlags::AGAINST_SUPERMAJORITY) {
+				Self::deposit_event(Event::DisputeConcluded(
+					set.candidate_hash,
+					DisputeResult::Invalid,
+				));
+			}
+		}
+
+		// Reward statements.
+		T::RewardValidators::reward_dispute_statement(
+			set.session,
+			summary.new_participants.iter_ones().map(|i| ValidatorIndex(i as _)),
+		);
+
+		// Slash participants on a losing side.
+		{
+			// a valid candidate, according to 2/3. Punish those on the 'against' side.
+			T::PunishValidators::punish_against_valid(
+				set.session,
+				summary.slash_against,
+			);
+
+			// an invalid candidate, according to 2/3. Punish those on the 'for' side.
+			T::PunishValidators::punish_for_invalid(
+				set.session,
+				summary.slash_for,
+			);
+		}
+
+		<Disputes<T>>::insert(&set.session, &set.candidate_hash, &summary.state);
+
+		// Freeze if just concluded against some local candidate
+		if summary.new_flags.contains(DisputeStateFlags::AGAINST_SUPERMAJORITY) {
+			if let Some(revert_to) = <Included<T>>::get(&set.session, &set.candidate_hash) {
+				Self::revert_and_freeze(revert_to);
```

Comment:
It's unlikely but possible that two different disputes would resolve at this block, each one asking to revert to different blocks. I assume the expected behavior is that the client reverts to the earliest one, would it make sense to deduplicate in the runtime?

Reply:
Yes we need reversions to merge.

Reply:
Right, good point.

Reply:
I added some code for merging reversions. FWIW, after the first reversion, the chain is already likely to be abandoned and any prior dispute should be transplanted, leading to yet another reversion.